ocaml/typing/ctype.ml

```diff
@@ -3653,6 +3941,8 @@ and unify3 env t1 t1' t2 t2' =
           raise_for Unify (Obj (Abstract_row First))
       | (_, _) -> raise_unexplained_for Unify
       end;
+      if in_functor then
+        Transient_expr.set_desc tt1' d1;        
```

Comment:
Here and in some other places you break the property that unification makes types physically equal -- so that future mutations of one input affect the other. This breaks OCaml's model of the world and, in particular, it's approach to ensuring principality.

For example, the following code is not principal, and triggers a principality warning upstream if you use `-principal`, but this branch gives no warning:
```ocaml
type float2 = float

let forty_two : float2 = 42.0

type 'a ty =
  | Float : float ty
  | String : string ty

let bar (type a) p (ty : a ty) (a : a) =
  match ty with
  | Float ->
      let r = ref None in
      r.contents <- Some forty_two;
      let bar () =
        match r.contents with
        | None -> 3.5
        | Some x -> x
      in
      r.contents <- Some a;
      bar ()
  | String -> assert false
```

I think you need to change your approach slightly so that the identifiers bounds by `Tfunctor` nodes are mutable such that one can be mutated to be permanently considered equal to another. This would allow you to have unification continue to make types physically equal.

I even wonder if it might be better for these identifiers to be a separate type from `Ident.t` with their own constructor in `Path.t` and their own table in `Env.t`.

Reply:
The problem of principality should now be solved due to the implementation of an unscoped type for identifiers that are unifiable.